neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
 
-    def next_to_execute(self) -> NextExecute:
-        return self._all_encoders, {"bs_outputs": self._bs_outputs}, {}
+        self._next_feed = [{} for _ in range(self._num_sessions)] \
+            # type: List[FeedDict]
 
-    def collect_results(self, results: List[Dict]) -> None:
-        if len(results) > 1:
-            raise ValueError("Beam search runner does not support ensembling.")
+        # If we are ensembling, we run only one step at a time
+        # Note, that in the first iteration, we only want to generate
+        # logprobs for ensembling, therefore we set max_steps = 0
+        if self._num_sessions > 1:
+            for fd in self._next_feed:
+                fd.update({self._decoder.max_steps: 0})
 
-        evaluated_bs = results[0]["bs_outputs"]
-        max_time = evaluated_bs.scores.shape[0]
+        self.result = None  # type: ExecutionResult
 
-        # pick the end of the hypothesis based on its rank
-        hyp_index = np.argpartition(
-            -evaluated_bs.scores[-1], self._rank - 1)[self._rank - 1]
-        bs_score = evaluated_bs.scores[-1][hyp_index]
+    def next_to_execute(self) -> NextExecute:
+        return (self._all_encoders,
+                {"bs_outputs": self._decoder.outputs},
+                self._next_feed)
 
-        # now backtrack
-        output_tokens = []  # type: List[str]
+    # pylint: disable=too-many-locals
+    def collect_results(self, results: List[Dict]) -> None:
+        # Recompute logits
+        # Only necessary when ensembling models
+        prev_logprobs = [res["bs_outputs"].last_search_state.prev_logprobs
+                         for res in results]
+
+        # Arithmetic mean
+        ens_logprobs = (scipy.misc.logsumexp(prev_logprobs, 0)
+                        - np.log(self._num_sessions))
+
+        # The last score is computed
+        bs_outputs = results[0]["bs_outputs"]
+        step = bs_outputs.last_dec_loop_state.step - 1
+
+        self.step += step
+        self.scores = np.append(
+            self.scores,
+            bs_outputs.last_search_step_output.scores[0:step],
+            axis=0)
+
+        self.parent_ids = np.append(
+            self.parent_ids,
+            bs_outputs.last_search_step_output.parent_ids[0:step],
+            axis=0)
+
+        self.token_ids = np.append(
+            self.token_ids,
+            bs_outputs.last_search_step_output.token_ids[0:step],
+            axis=0)
+
+        if (self._decoder.max_output_len is not None and
+                self.step > self._decoder.max_output_len):
+            self.prepare_results()
+            return
+
+        # Prepare the next feed_dict (in ensembles)
+        self._next_feed = []
+        for result in results:
+            bs_outputs = result["bs_outputs"]
+
+            search_state = bs_outputs.last_search_state._replace(
+                prev_logprobs=ens_logprobs)
+
+            fd = {self._decoder.max_steps: 1,
```

Comment:
a tady už 1 je, tak teď tomu ještě víc nerozumim

Reply:
Opet jde o to, jak se vola beamsearch decoder.
1. spustim decoder_body
2. priradim do attributu beamsearch_decoder objektu vystupni tensory z decoder_body, aby se daly v budoucnu feedovat
3. pousti se tf.while

V prvnim kroku dekodingu se mi vytvori logproby uz behem 1. a ty chci ensemblovat; max_len zpusobi, ze se 3. vubec nezavola.
V dalsich krocich uz se decoder_body z 1. nevola, protoze se ty tensory nafeeduji do attributu vzniklych v 2. a tedy chci pustit jeden krok z tf.while (3.) abych vybral vystupni symboly na zaklade jiz ensemblovanych logprobu + vygeneroval nove logproby do dalsicho kroku.

Reply:
Kdyz vymyslis chytrejsi zpusob, jak to delat, budu rad :)

Reply:
nevadí mi ten kód jako takovej, jen bych tam chtěl mít komentáře. neni to žádnej kód, kterej zapadne, tohle se bude používat. takže to musí bejt pořádně pochopitelný, co se děje, aby to šlo udržovat.

Reply:
S tim souhlasim. Dodelam rovnou i docstringy s hrubym popisem procesu v beamsearch_runner.py i beamsearch_decoder.py. Akorat nevim, jestli to bude jeste dneska.

Reply:
IN PROGRESS